code/kobe.py

```diff
@@ -1107,32 +1117,543 @@ def kobe_shadows(df_input,cdpp):
+    delta_f = np.arcsin(exp)
+    t1 = df_kobe_output[col_period]/np.pi
+    t2 = rho_c**2/(np.sqrt(1 - df_kobe_output[col_ecc]*df_kobe_output[col_ecc]))
+   
+    transit_duration = t1*t2*delta_f*24
+    
+    return transit_duration
+
+def kobe_detection_efficiency(df_kobe_output, detection_efficiency):
+
+    """
+    Defined with KOBE version 2.0
+    
+    This function marks the planets from KOBE Shadows catalogue as TCEs or not TCEs.
+    To perform this procedure, the 1D pipeline detection efficiency as defined in Planet Detection Metrics by
+    Jessie L. Christiansen 2017, is used.
+
+    NOTE
+    ----
+    1. The detection efficiency is calculated using a gamma cumulative distribution function, values of the 
+       coefficients taken directly from Christiansen 2017.
+       
+    2. Flag 'flag_det_eff' is created
+       '1' for planets which are tces
+       '-1' for planets which are not tces
+    
+    3. Detection efficiency is solely a function of the MES. For MES values greater than 50, all planets within
+       a bin are assigned a detection efficiency of 100%. Gamma CDF is used for MES ranging between 0 and 50.
+       This value of 50 can be changed in the function to define your bins differently (max_bin). 
+    
+    Parameters
+    ----------
+    df_kobe_output : pandas dataframe
+        dataframe of planets that transit (potentially detectable)
+        columns: all from input, some are added
+    
+    detection_efficiency : a boolean input to be specified in the file kobe_choices
+         set to True for using the detection efficiency for marking TCEs
+         
+    Output 
+    ------
+    df_kobe_output : pandas dataframe
+        dataframe of all KOBE Shadows planets that can be recovered by the Kepler pipeline 
+        as function of their MES, flagged as TCEs
+        columns: all from input, some are added
+    
+    """
+    # Two flags are created. 
+    # 'flag_det_eff' : '1' means planet is marked as a TCE. '-1' means planet is not marked as a TCE.
+    
+    ## 0 - create arrays to go through each bin       
+    ## 1 - sanity checks
+    drop_c=0
+    drop1_c=0
+    
+    ## Create Columns 
+    ## 1 create columns and mark each as '1'. This ensures that all potential planets with MES>50 are marked TCEs.
+    df_kobe_output['flag_det_eff'] = 1
+    ## then all planets that are not tce are marked with -1
+    
+    ## select bin width, the MES threshold after which potential planets are marked TCEs
+    min_bin = 0
+    max_bin = 50
+    bin_wid = 0.5
+    n_bins = ((max_bin - min_bin)/bin_wid)
+    
+    for i in np.arange(min_bin, max_bin, bin_wid):
+        min_mes = i
+        max_mes = min_mes + 0.5
+        
+        #select all planets that fall within the bin (min_mes, max_mes)
+        df_bin = df_kobe_output.loc[df_kobe_output['kobe_mes']>=min_mes].loc[df_kobe_output['kobe_mes']<max_mes]    
+        
+        for mes in df_bin['kobe_mes']:
```

Comment:
we enter this loop only to get the average mes? Can we not used the middle of the bin and avoid going in the loop, somehow?